accepted/tail-calls.md

```diff
@@ -0,0 +1,1239 @@
+
+* It does introduce some new memory overhead, since we grow the size of
+  `VMCallerCheckedAnyfunc` from three words to five words. This is a fundamental
+  space-time trade off, where our current design (unwittingly) trades time for
+  space by chaining together trampolines instead of keeping around multiple
+  function pointers, each for use by a different kind of caller. I believe that
+  this space overhead is overall acceptable since `VMCallerCheckedAnyfunc`s
+  account for a small fraction of the total space needed for a Wasm instance,
+  and not even all Wasm functions need to be materialized as
+  `VMCallerCheckedAnyfunc`s in the first place.
+
+* Multiple copies of identical trampolines across different compiled Wasm
+  modules are not deduplicated. That deduplication is not feasible unless we
+  give up AOT compilation, and instead JIT compile trampolines the first time we
+  actually need them and haven't already JIT compiled that particular
+  trampoline. This seems like an unavoidable cost given that relying on JIT
+  compilation is a nonstarter.
+
+That said, it still puts us in a better place than any of the alternative
+trampoline designs:
+
+* Require a compiler to always be built into Wasmtime so we can JIT trampolines
+  on demand to make certain function calls. But requiring JIT compilation is a
+  non-starter.
+
+* Dynamically detect when we would need a trampoline but don't have one (i.e. we
+  would need JIT compilation) and panic in these scenarios. Even if rare, this
+  is very opaque and hard to understand for Wasmtime embedders unless they know
+  the details of Wasmtime's implementation.
+
+* Dynamically inspect the callee and caller pair, and branch to different code
+  paths depending on whether it is Wasm that needs a trampoline or is a
+  host-to-host call that doesn't need trampolines. This adds complexity and
+  overhead to our call paths. We want these to be as maintainable as possible
+  and definitely as fast as possible.
+
+* Add Cranelift as a build dependency for Wasmtime and pre-compile a trampoline
+  for each of our `IntoFunc` signatures so we always have trampolines available
+  for host functions. In addition to complicating the build process, this bloats
+  the binary with a bunch of trampolines that are probably never used, which
+  makes it harder to use Wasmtime in resource constrained environments.
+
+* Use the type system to differentiate between `wasmtime::HostFunc` and
+  `wasmtime::WasmFunc` and then statically disallow calls that would need some
+  sort of trampoline that we wouldn't have access to. This is a large, clunky
+  change to Wasmtime's API for very little benefit, especially since we have
+  viable alternatives.
+
+As far as the new Wasm calling convention goes, there are some alternative
+approaches that we could explore as well:
+
+* We could bend over backwards to maintain interoperability with existing
+  calling conventions (to a degree) the way that [SpiderMonkey plans on
```

Comment:
Our plans are still in flux here, we're currently prototyping out some other approaches other than what Lars prototyped in that document.

For extra context too, our primary complexity here is that our ABI has some global state (pinned registers and some JS::Activation state) that must be kept in sync with our current module instance. Upon calls that might cross module instances (imports/indirect/call_ref), the caller performs a context switch before and after the call to maintain this global state. We don't do this work on direct calls, as we statically whether it will be cross-instance. This biases us in favor of direct calls, as they're assumed to be the critical path.

With tail calls however, we can replace a same-instance callee with a cross-instance callee and when that cross-instance callee returns we need to do a context-switch, but the direct caller is not set up to do that.

Lars' scheme adds an extra frame on the first tail call to prevent this from happening. This has the noted complexity, so we're evaluating some other options too.

Reply:
Good to know, thanks @eqrion. Interested to hear where y'all end up in the design space.